packages/preferences/src/browser/abstract-resource-preference-provider.ts

```diff
@@ -15,59 +15,80 @@
 import URI from '@theia/core/lib/common/uri';
 import { PreferenceConfigurations } from '@theia/core/lib/browser/preferences/preference-configurations';
+import { MonacoTextModelService } from '@theia/monaco/lib/browser/monaco-text-model-service';
+import { MonacoEditorModel } from '@theia/monaco/lib/browser/monaco-editor-model';
+import { MonacoWorkspace } from '@theia/monaco/lib/browser/monaco-workspace';
+import { Deferred } from '@theia/core/lib/common/promise-util';
 
 @injectable()
 export abstract class AbstractResourcePreferenceProvider extends PreferenceProvider {
 
     protected preferences: { [key: string]: any } = {};
-    protected resource: Promise<Resource>;
+    protected model: MonacoEditorModel | undefined;
+    protected readonly loading = new Deferred();
 
     @inject(PreferenceService) protected readonly preferenceService: PreferenceService;
-    @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider;
     @inject(MessageService) protected readonly messageService: MessageService;
     @inject(PreferenceSchemaProvider) protected readonly schemaProvider: PreferenceSchemaProvider;
 
     @inject(PreferenceConfigurations)
     protected readonly configurations: PreferenceConfigurations;
 
+    @inject(MonacoTextModelService)
+    protected readonly textModelService: MonacoTextModelService;
+
+    @inject(MonacoWorkspace)
+    protected readonly workspace: MonacoWorkspace;
+
     @postConstruct()
     protected async init(): Promise<void> {
         const uri = this.getUri();
-        this.resource = this.resourceProvider(uri);
+        this.toDispose.push(Disposable.create(() => this.loading.reject(new Error(`preference provider for '${uri}' was disposed`))));
+        this._ready.resolve();
 
-        // Try to read the initial content of the preferences.  The provider
-        // becomes ready even if we fail reading the preferences, so we don't
-        // hang the preference service.
-        this.readPreferences()
-            .then(() => this._ready.resolve())
-            .catch(() => this._ready.resolve());
-
-        const resource = await this.resource;
-        this.toDispose.push(resource);
-        if (resource.onDidChangeContents) {
-            this.toDispose.push(resource.onDidChangeContents(() => this.readPreferences()));
+        const reference = await this.textModelService.createModelReference(uri);
+        if (this.toDispose.disposed) {
+            reference.dispose();
+            return;
         }
+
+        this.model = reference.object;
+        this.loading.resolve();
+
+        this.toDispose.push(reference);
+        this.toDispose.push(Disposable.create(() => this.model = undefined));
+
+        this.readPreferences();
+        this.toDispose.push(this.model.onDidChangeContent(() => this.readPreferences()));
+        this.toDispose.push(this.model.onDirtyChanged(() => this.readPreferences()));
```

Comment:
would it make sense to debounce `readPreferences` calls?

Reply:
It should guruante that preferences are reloaded if monaco model is changed on save. If we debounce it won't.

Reply:
discussed that there is no inconsistency issues, since `readPreferences` is sync and can be executed only one after another